### PR TITLE
Add store to protocol agents

### PR DESCRIPTION
**Pull request by zhang-zichao**

### CODE REVIEW

File: Taglierino.hs [hunk 1]
```diff
@@ -345,6 +345,26 @@ receive = do
...
+insertFreshL :: [LTS.LabelPart] -> LTS.Label
+insertFreshL args = LTS.Label $ [LTS.Simple "insertfresh"] ++ args
+
+insertFresh :: Term -> Proc Bool
+insertFresh r = do
+  ProcRState{..} <- ask
+  case S.lookupIndex r psAllowed of
+    Just mi -> 
+      Proc $ ContT $ \k -> do
+        s <- get
```
Comment (arthuraa) [inline Taglierino.hs:357]: Why are you reading the state and writing it back?

File: Taglierino.hs [hunk 2]
```diff
@@ -910,12 +945,95 @@ honestRange agents =
...
+    combinations 0 _ = [[]]
+    combinations n xs = [y : ys | y:xs' <- tails xs, ys <- combinations (n - 1) xs']
+    
+    allComb :: Ord a => [a] -> [S.Set a]
+    allComb l = map S.fromList $ concat result
+      where len = length l
+            result = map (`combinations`  l) [0.. len]
+
+    entries :: [S.Set Term]
+    entries = allComb $ S.toList records
```
Comment (arthuraa) [inline Taglierino.hs:969]: I think `Data.Set.powerSet` does exactly this.

File: Taglierino.hs [hunk 3]
```diff
@@ -949,7 +1067,9 @@ compileWith opts@Options{..} sys =
           hPrint h $ pretty def
         hPutStrLn h "// Attacker"
         hPrint h $ pretty $ attacker opts pPublic pAllowed pKnowledgeSize
-        hPrint h $ pretty $ LTS.Parallel "System" $ LTS.Primitive $ "Attacker" : map (LTS.name . fst) (M.elems compiledAgents)
+        hPutStrLn h "// Storages"
+        forM_ (M.toList pStore) (\(name, records)-> hPrint h $ pretty $ makeStore pAllowed records name (numInstance name pProcs)) 
+        hPrint h $ pretty $ LTS.Parallel "System" $ LTS.Primitive $ ("Attacker" : map (LTS.name . fst) (M.elems compiledAgents)) ++ map ("Store_" ++) (M.keys pStore)
```
Comment (arthuraa) [inline Taglierino.hs]: We could split this expression using intermediate variables, so that it is easier to read.

File: Taglierino.hs [hunk 4]
```diff
@@ -345,6 +345,26 @@ receive = do
...
+        s <- get
+        actions <- forM [True, False] $ \m -> do
+          put s
+          body <- k m
+          let mm = if m then "true" else "false"
+          let l = insertFreshL $ [agentLabel psAgent, LTS.const psThread] ++
+                  labelOfTerm psOptions r mi ++ [LTS.Simple mm]
+          return $ LTS.Action l body
+        return $ LTS.Body actions
+    Nothing -> fail $ show $ pretty "Not allowed to insertfresh" <+> pretty r <+> pretty " in " <+> pretty (LTS.Simple (agentName psAgent)) <+> pretty "'s storage"
```
Comment (arthuraa) [inline Taglierino.hs]: Break this line.

File: Taglierino.hs [hunk 5]
```diff
@@ -571,6 +603,7 @@ data Program =
   Program { pEvents        :: S.Set Term
           , pAllowed       :: S.Set Term
           , pPublic        :: S.Set Term
+          , pStore         :: M.Map String (S.Set Term)
```
Comment (arthuraa) [inline Taglierino.hs:609]: If the store is part of the agent declaration, maybe we could create a new `AgentDef` data type for holding the agent's code _and_ its table.

File: Taglierino.hs [hunk 6]
```diff
@@ -910,12 +945,95 @@ honestRange agents =
...
+
+    initialState = stateLookup []
+
+    serializeRecords :: Term -> Int
+    serializeRecords r = S.findIndex r allowed 
+
+    oneState :: S.Set Term -> [Int] -> [LTS.Action]
+    oneState s ns =
+      let  rest = S.difference records s
+           ss = (\x -> LTS.Label [LTS.Variable (show x)]) <$> ns
```
Comment (arthuraa) [inline Taglierino.hs]: This would be clearer with a list comprehension `[LTS.Label ... | x <- ns]`

File: Taglierino.hs [hunk 7]
```diff
@@ -910,12 +945,95 @@ honestRange agents =
...
+           actionT m = LTS.Action (insertFreshL$ [LTS.Simple name, LTS.Set ss, LTS.Variable (show (serializeRecords m)), LTS.Simple "true"]) (stateLookup (s `S.union` (S.singleton m)))
+           fs = (\x -> LTS.Label [LTS.Variable (show (serializeRecords x))]) <$> S.toList s
```
Comment (arthuraa) [inline Taglierino.hs]: Ditto; use a list comprehension.

File: Taglierino.hs [hunk 8]
```diff
@@ -910,12 +945,95 @@ honestRange agents =
...
+    initialState = stateLookup []
+
+    serializeRecords :: Term -> Int
+    serializeRecords r = S.findIndex r allowed 
+
+    oneState :: S.Set Term -> [Int] -> [LTS.Action]
+    oneState s ns =
+      let  rest = S.difference records s
+           ss = (\x -> LTS.Label [LTS.Variable (show x)]) <$> ns
+           actionT m = LTS.Action (insertFreshL$ [LTS.Simple name, LTS.Set ss, LTS.Variable (show (serializeRecords m)), LTS.Simple "true"]) (stateLookup (s `S.union` (S.singleton m)))
```
Comment (arthuraa) [inline Taglierino.hs]: The definitions of `actionT` and `actionF` are hard to read right now.

**Review by arthuraa — CHANGES_REQUESTED** [review]

File: Taglierino.hs (hunk 2 above)
Comment (zhang-zichao) [reply Taglierino.hs:969]: I add powerset but commented it out to keep the generated model the same.

File: Taglierino.hs (hunk 5 above)
Comment (zhang-zichao) [reply Taglierino.hs:609]: Yes, this would be better. Maybe I could it in the future.

File: Taglierino.hs (hunk 1 above)
Comment (zhang-zichao) [reply Taglierino.hs:357]: I tried to keep it same as receive since insertFresh is kind of like receive where it branches to different state.

**Review by arthuraa — APPROVED** [review]

**Pull request merged**